File: PackIt/src/kotlin/sh/packit/core/utils/Logx.kt
package sh.packit.core.utils

import android.content.Context
import android.util.Log
import de.shareui.exterasdk.utils.AndroidUtils
import org.json.JSONObject
import org.telegram.messenger.ApplicationLoader
import java.io.File
import java.io.FileOutputStream
import java.nio.channels.FileLock
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object Logx {
    private const val TAG: String = "PackIt"
    private val writeLock: Any = Any()
    @Volatile
    private var activeLogFile: File? = null
    @Volatile
    var isDebugEnabled: Boolean = true
    @Volatile
    var isWriteLogsEnabled: Boolean = false

    init {
        reloadConfig()
    }

    private fun getContext(): Context? {
        return try {
            ApplicationLoader.applicationContext
        } catch (e: Throwable) {
            Log.d(TAG, "failed to get ApplicationLoader context: $e")
            null
        }
    }

    private fun getFilesDir(): String? {
        val ctx: Context = getContext() ?: return null
        return ctx.filesDir.absolutePath
    }

    private fun getPluginsDir(): String? {
        val filesDir: String = getFilesDir() ?: return null
        return "$filesDir/plugins"
    }

    private fun getLogsDir(): File? {
        val filesDir: String = getFilesDir() ?: return null
        return File(filesDir, "packit/var/logs")
    }

    @JvmStatic
    fun initLogSession(force: Boolean = false): File? {
        if (!force) {
            activeLogFile?.let { return it }
        }
        val logsDir: File = getLogsDir() ?: return null
        val historyDir = File(logsDir, "history")
        if (!historyDir.exists()) {
            historyDir.mkdirs()
        }
        val fileName: String = SimpleDateFormat("yy:MM:dd-HH:mm:ss", Locale.US).format(Date()) + ".txt"
        val sessionFile = File(historyDir, fileName)
        try {
            if (!sessionFile.exists()) {
                sessionFile.createNewFile()
            }
        } catch (e: Throwable) {
            AndroidUtils.log("[packit] failed to create history file: $e")
        }
        val latestFile = File(logsDir, "latest.txt")
        try {
            android.system.Os.remove(latestFile.absolutePath)
        } catch (e: Throwable) {
            Log.d(TAG, "remove old latest.txt notice: $e")
        }
        try {
            android.system.Os.symlink(sessionFile.absolutePath, latestFile.absolutePath)
        } catch (e: Throwable) {
            AndroidUtils.log("[packit] failed to symlink latest.txt: $e")
        }
        val resolved = if (latestFile.exists()) latestFile else sessionFile
        activeLogFile = resolved
        return resolved
    }

    private fun getLogFile(): File? {
        activeLogFile?.let { return it }
        val logsDir: File = getLogsDir() ?: return null
        val latestFile = File(logsDir, "latest.txt")
        if (latestFile.exists()) {
            activeLogFile = latestFile
            return latestFile
        }
        return initLogSession()
    }

    @JvmStatic
    fun reloadConfig() {
        val pluginsDir: String = getPluginsDir() ?: return
        try {
            val file = File(pluginsDir, "plugin_settings.json")
            if (!file.exists()) {
                isDebugEnabled = true
                isWriteLogsEnabled = false
                return
            }
            val json = JSONObject(file.readText())
            val pluginObj: JSONObject? = json.optJSONObject("packit") ?: json.optJSONObject("shareui_packit")
            isDebugEnabled = pluginObj?.optBoolean("debug_logs", true) ?: true
            isWriteLogsEnabled = pluginObj?.optBoolean("write_logs", false) ?: false
        } catch (e: Throwable) {
            AndroidUtils.log("[packit] failed to read settings in reloadConfig: $e")
        }
    }

    @JvmStatic
    fun setDebugLogs(enabled: Boolean) {
        isDebugEnabled = enabled
    }

    @JvmStatic
    fun setWriteLogs(enabled: Boolean) {
        isWriteLogsEnabled = enabled
    }

    private fun writeToFile(msg: String) {
        val file: File = getLogFile() ?: return
        val ts: String = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US).format(Date())
        val fullLine: String = "$ts ~> $msg\n"
        val bytes: ByteArray = fullLine.toByteArray(Charsets.UTF_8)
        synchronized(writeLock) {
            try {
                FileOutputStream(file, true).use { fos ->
                    val channel = fos.channel
                    var fileLock: FileLock? = null
                    try {
                        fileLock = channel.lock()
                    } catch (e: Throwable) {
                        Log.d(TAG, "channel.lock notice: $e")
                    }
                    try {
                        fos.write(bytes)
                        fos.flush()
                    } finally {
                        try {
                            fileLock?.release()
                        } catch (e: Throwable) {
                            Log.d(TAG, "channel lock release notice: $e")
                        }
                    }
                }
            } catch (e: Throwable) {
                AndroidUtils.log("[packit] failed to write log: $e")
            }
        }
    }

    @JvmStatic
    fun logx(msg: String, isDebug: Boolean = false) {
        val formatted: String = if (msg.startsWith("[packit]")) msg else "[packit] $msg"
        if (isDebug && !isDebugEnabled) {
            return
        }
        try {
            AndroidUtils.log(formatted)
        } catch (e: Throwable) {
            try {
                val fileLogCls: Class<*> = Class.forName("org.telegram.messenger.FileLog")
                fileLogCls.getMethod("d", String::class.java).invoke(null, formatted)
            } catch (_: Throwable) {
            }
            Log.d(TAG, "AndroidUtils.log fallback: $e")
            Log.d(TAG, formatted)
        }
        if (isWriteLogsEnabled) {
            writeToFile(formatted)
        }
    }
}

File: PackIt/src/wheels/packutil/package/logx.py
# pyright: reportMissingImports=false

import fcntl
import json
import os
import threading
import time
from android_utils import log as _log

_writeLock = threading.Lock()
_activeLogPath: str | None = None

def _getFilesDir() -> str:
    try:
        from org.telegram.messenger import ApplicationLoader
        return ApplicationLoader.applicationContext.getFilesDir().getAbsolutePath()
    except Exception as e:
        _log(f"[packit] failed to get filesDir: {e}")
        return "/data/data/org.telegram.messenger.web/files"

def _getPluginsDir() -> str:
    return _getFilesDir() + "/plugins"

def _getLogsDir() -> str:
    return _getFilesDir() + "/packit/var/logs"

def _getHistoryDir() -> str:
    return _getLogsDir() + "/history"

def initLogSession(force: bool = False) -> str:
    global _activeLogPath
    if _activeLogPath is not None and not force:
        return _activeLogPath
    logsDir = _getLogsDir()
    historyDir = _getHistoryDir()
    try:
        os.makedirs(historyDir, exist_ok=True)
    except Exception as e:
        _log(f"[packit] failed to create history dir: {e}")
    fileName = time.strftime("%y:%m:%d-%H:%M:%S.txt")
    historyFile = os.path.join(historyDir, fileName)
    try:
        if not os.path.exists(historyFile):
            open(historyFile, "a", encoding="utf-8").close()
    except Exception as e:
        _log(f"[packit] failed to touch history file: {e}")
    latestFile = os.path.join(logsDir, "latest.txt")
    try:
        if os.path.lexists(latestFile):
            os.unlink(latestFile)
        os.symlink(historyFile, latestFile)
    except Exception as e:
        _log(f"[packit] failed to symlink latest.txt: {e}")
    _activeLogPath = latestFile if os.path.exists(latestFile) else historyFile
    return _activeLogPath

_isDebugEnabled: bool = True
_isWriteLogsEnabled: bool = False

def reloadConfig() -> None:
    global _isDebugEnabled, _isWriteLogsEnabled
    try:
        from elyx import settings as _s
        val_debug = _s.get("debug_logs", None)
        val_write = _s.get("write_logs", None)
        if val_debug is not None or val_write is not None:
            if val_debug is not None:
                _isDebugEnabled = bool(val_debug)
            if val_write is not None:
                _isWriteLogsEnabled = bool(val_write)
            return
    except Exception:
        pass
    try:
        path = _getPluginsDir() + "/plugin_settings.json"
        if not os.path.exists(path):
            return
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
        pdata = data.get("packit") or data.get("shareui_packit") or {}
        _isDebugEnabled = bool(pdata.get("debug_logs", True))
        _isWriteLogsEnabled = bool(pdata.get("write_logs", False))
    except Exception as e:
        _log(f"[packit] reloadConfig failed: {e}")

def setDebugLogs(enabled: bool) -> None:
    global _isDebugEnabled
    _isDebugEnabled = bool(enabled)

def setWriteLogs(enabled: bool) -> None:
    global _isWriteLogsEnabled
    _isWriteLogsEnabled = bool(enabled)

def _writeToFile(msg: str):
    global _activeLogPath
    path = _activeLogPath
    if path is None:
        path = initLogSession()
    if path is None:
        return
    ts = time.strftime("%Y-%m-%d %H:%M:%S")
    fullLine = f"{ts} ~> {msg}\n"
    with _writeLock:
        try:
            with open(path, "a", encoding="utf-8") as f:
                try:
                    fcntl.lockf(f.fileno(), fcntl.LOCK_EX)
                except Exception as e:
                    _log(f"[packit] failed to acquire lockf: {e}")
                try:
                    f.write(fullLine)
                    f.flush()
                finally:
                    try:
                        fcntl.lockf(f.fileno(), fcntl.LOCK_UN)
                    except Exception as e:
                        _log(f"[packit] failed to release lockf: {e}")
        except Exception as e:
            _log(f"[packit] failed to write log: {e}")

def logx(msg: str, isDebug: bool):
    formatted = str(msg)
    if not formatted.startswith("[packit]"):
        formatted = f"[packit] {formatted}"
    if isDebug and not _isDebugEnabled:
        return
    _log(formatted)
    if _isWriteLogsEnabled:
        _writeToFile(formatted)

reloadConfig()
